### PR TITLE
feat: Add KSpinner (issue #118)

### PR DESCRIPTION
**Pull request by michaelbukachi**
This pull request adds Kspinner as discussed in  issue #118 and pull request #119.
@Unlimity do you mind reviewing the code. I still want to add more assertions such as `isOpened()`, 
`isClosed()`

### CODE REVIEW

**Comment by Unlimity** [comment]
Hi!
Thanks for contribution. Right now I'm in a middle of a trip with limited time and access to laptop and will be able to review your PR next week only.
Sorry for inconvenience.

**Comment by michaelbukachi** [comment]
It's alright. No problem.

**Comment by Vacxe** [comment]
@michaelbukachi @Unlimity I will try to review it today evening

**Comment by bvdaakster** [comment]
Would appreciate to have this merged as well

File: kakao/src/main/kotlin/com/agoda/kakao/spinner/KSpinner.kt [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+     * @param function Tail lambda which receiver will be matched item with given type T
+     */
+    inline fun <reified T : KAdapterItem<*>> childAt(position: Int, function: T.() -> Unit) {
+        val provideItem = itemTypes.getOrElse(T::class) {
+            throw IllegalStateException("${T::class.java.simpleName} did not register to Spinner")
+        }.provideItem
+
+        val interaction = Espresso.onData(Matchers.anything())
+            .inRoot(isPlatformPopup())
+//            .inAdapterView(matcher)
```
Comment (Unlimity) [inline kakao/src/main/kotlin/com/agoda/kakao/spinner/KSpinner.kt]: It seems this code is unnecessary. Although, to reduce the chance of multiple matching exception, you can bind this interaction to the root view class of the popup as well.

File: kakao/src/main/kotlin/com/agoda/kakao/spinner/KSpinner.kt [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    override fun open() {
+        view.perform(click())
+        popupView = ViewInteractionDelegate(onView(allOf(ViewMatchers.isAssignableFrom(AdapterView::class.java))).inRoot(isPlatformPopup()))
+    }
+
+    override fun close() {
+        view.perform(click())
+        popupView = null
+    }
+}
```
Comment (Unlimity) [inline kakao/src/main/kotlin/com/agoda/kakao/spinner/KSpinner.kt]: New line missing

File: kakao/src/main/kotlin/com/agoda/kakao/spinner/SpinnerAdapterActions.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    /**
+     * Opens the spinner to display it's items
+     */
+    fun open()
+
+    /**
+     * Closes the spinner
+     */
+    fun close()
+}
```
Comment (Unlimity) [inline kakao/src/main/kotlin/com/agoda/kakao/spinner/SpinnerAdapterActions.kt]: New line missing

File: kakao/src/main/kotlin/com/agoda/kakao/spinner/SpinnerAdapterAssertions.kt [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        view.check(ViewAssertions.matches(SpinnerAdapterSizeMatcher(size)))
+    }
+
+
+    /**
+     * Check selected text
+     *
+     * @param text expected selected text in Spinner
+     */
+    fun withText(text: String) {
```
Comment (Unlimity) [inline kakao/src/main/kotlin/com/agoda/kakao/spinner/SpinnerAdapterAssertions.kt]: I think `hasText` is better naming for assertion

File: kakao/src/main/kotlin/com/agoda/kakao/spinner/SpinnerAdapterAssertions.kt [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        view.check(ViewAssertions.matches(withSpinnerText(containsString(text))))
+    }
+}
```
Comment (Unlimity) [inline kakao/src/main/kotlin/com/agoda/kakao/spinner/SpinnerAdapterAssertions.kt]: New line missing

File: sample/src/main/AndroidManifest.xml [hunk 6]
```diff
@@ -3,74 +3,64 @@
...
-
     <activity
         android:name=".AlertDialogActivity"
         android:label="Alert Dialog Activity"
         android:theme="@style/AppTheme" />
-
   </application>
 
-</manifest>
+</manifest>
```
Comment (Unlimity) [inline sample/src/main/AndroidManifest.xml]: New line missing

File: sample/src/androidTest/kotlin/com/agoda/sample/screen/SpinnerActivityScreen.kt [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+import com.agoda.kakao.spinner.KSpinnerItem
+import com.agoda.sample.R
+
+class SpinnerActivityScreen : Screen<SpinnerActivityScreen>() {
+    val spinner = KSpinner(
+        builder = { withId(R.id.spinner) },
+        itemTypeBuilder = { itemType(::KSpinnerItem) }
+    )
+
+}
```
Comment (Unlimity) [inline sample/src/androidTest/kotlin/com/agoda/sample/screen/SpinnerActivityScreen.kt]: New line missing

File: sample/src/androidTest/kotlin/com/agoda/sample/SpinnerTest.kt [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+                    isDisplayed()
+                    hasText("Title 5")
+                    click()
+                }
+
+                withText("Title 5")
+            }
+        }
+    }
+}
```
Comment (Unlimity) [inline sample/src/androidTest/kotlin/com/agoda/sample/SpinnerTest.kt]: New line missing

**Comment by Vacxe** [comment]
@michaelbukachi you can easily add new lines, to your code. Just check script in contribution policy. Also don't forget to apply code style to this part of the code.

**Comment by Vacxe** [comment]
Please polish the code and we can merge it. 

Cheers!

File: kakao/src/main/kotlin/com/agoda/kakao/spinner/SpinnerAdapterActions.kt [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+                    size = view.count
+                }
+            }
+
+        })
+
+        return size
+    }
+
+
```
Comment (Vacxe) [inline kakao/src/main/kotlin/com/agoda/kakao/spinner/SpinnerAdapterActions.kt]: And also double empty lines

File: kakao/src/main/kotlin/com/agoda/kakao/spinner/SpinnerAdapterActions.kt [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+            override fun getDescription() = "Get AdapterView adapter size"
+
+            override fun getConstraints() = Matchers.allOf(ViewMatchers.isAssignableFrom(AdapterView::class.java), ViewMatchers.isDisplayed())
+
+            override fun perform(uiController: UiController?, view: View?) {
+                if (view is AdapterView<*>) {
+                    size = view.count
+                }
+            }
+
```
Comment (Vacxe) [inline kakao/src/main/kotlin/com/agoda/kakao/spinner/SpinnerAdapterActions.kt]: Not necessary line

File: kakao/src/main/kotlin/com/agoda/kakao/spinner/KSpinner.kt (hunk 1 above)
Comment (michaelbukachi) [reply kakao/src/main/kotlin/com/agoda/kakao/spinner/KSpinner.kt]: I haven't understood this. Isn't `isPlatformPopup` already doing the binding?

**Comment by michaelbukachi** [comment]
@Vacxe I've corrected the newlines issue

File: kakao/src/main/kotlin/com/agoda/kakao/spinner/KSpinner.kt (hunk 1 above)
Comment (Unlimity) [reply kakao/src/main/kotlin/com/agoda/kakao/spinner/KSpinner.kt]: That is the root matcher, basically it matches the windows inside the windows stack of your app. But you might have more than 1 windows opened alongside your main content's window at the same time.
In that situation you will get multiple matching exception.
So, if it's possible to distinctively detect the root view of a popup window, we should do that.

Comment (michaelbukachi) [reply kakao/src/main/kotlin/com/agoda/kakao/spinner/KSpinner.kt]: I see. The only way I can think of distinctively identifying popup root view is checking if's a descendant of of the DropDownView. Do you have a better constraint I should implement?

**Comment by Vacxe** [comment]
@michaelbukachi that's great, but could you also fix all comments placed by @Unlimity and me?

File: kakao/src/main/kotlin/com/agoda/kakao/spinner/KSpinner.kt (hunk 1 above)
Comment (Unlimity) [reply kakao/src/main/kotlin/com/agoda/kakao/spinner/KSpinner.kt]: I think that should do cover most of the cases.

Comment (michaelbukachi) [reply kakao/src/main/kotlin/com/agoda/kakao/spinner/KSpinner.kt]: @Unlimity I've added the bounds.

File: kakao/src/main/kotlin/com/agoda/kakao/spinner/helpers.kt [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+import androidx.test.espresso.matcher.RootMatchers.withDecorView
+import androidx.test.espresso.matcher.ViewMatchers.hasDescendant
+import androidx.test.espresso.matcher.ViewMatchers.withClassName
+import org.hamcrest.CoreMatchers.`is`
+import org.hamcrest.CoreMatchers.allOf
+import org.hamcrest.Description
+import org.hamcrest.Matcher
+import org.hamcrest.TypeSafeMatcher
+
+fun isSpinnerPopup(): Matcher<Root> = SpinnerPopup()
```
Comment (Vacxe) [inline kakao/src/main/kotlin/com/agoda/kakao/spinner/helpers.kt]: Could you follow the previous implementation convention for TypeSafeMatcher usage? We don't appreciate to use static functions in our code. 

`isSpinnerPopup` should be a part of Assertion  interface

File: kakao/src/main/kotlin/com/agoda/kakao/spinner/helpers.kt [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+
```
Comment (Vacxe) [inline kakao/src/main/kotlin/com/agoda/kakao/spinner/helpers.kt]: Double empty lines

File: kakao/src/main/kotlin/com/agoda/kakao/spinner/SpinnerAdapterAssertions.kt [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    /**
+     * Check size of Spinner
+     *
+     * @param size expected child count size in Spinner
+     */
+    fun hasSize(size: Int) {
+        view.check(ViewAssertions.matches(SpinnerAdapterSizeMatcher(size)))
+    }
+
+
```
Comment (Vacxe) [inline kakao/src/main/kotlin/com/agoda/kakao/spinner/SpinnerAdapterAssertions.kt]: Double empty lines

File: kakao/src/main/kotlin/com/agoda/kakao/spinner/helpers.kt [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+        }
+        val dropdownClassName = "android.widget.DropDownListView"
+        return withDecorView(
+            allOf(
+                withClassName(`is`<String>(popupClassName)),
+                hasDescendant(withClassName(`is`<String>(dropdownClassName)))
+            )
+        ).matches(item)
+    }
+
```
Comment (Vacxe) [inline kakao/src/main/kotlin/com/agoda/kakao/spinner/helpers.kt]: This empty is not necessary

File: sample/src/androidTest/kotlin/com/agoda/sample/screen/SpinnerActivityScreen.kt [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+import com.agoda.kakao.spinner.KSpinner
+import com.agoda.kakao.spinner.KSpinnerItem
+import com.agoda.sample.R
+
+class SpinnerActivityScreen : Screen<SpinnerActivityScreen>() {
+    val spinner = KSpinner(
+        builder = { withId(R.id.spinner) },
+        itemTypeBuilder = { itemType(::KSpinnerItem) }
+    )
+
```
Comment (Vacxe) [inline sample/src/androidTest/kotlin/com/agoda/sample/screen/SpinnerActivityScreen.kt]: Remove this empty

File: kakao/src/main/kotlin/com/agoda/kakao/spinner/helpers.kt (hunk 11 above)
Comment (michaelbukachi) [reply kakao/src/main/kotlin/com/agoda/kakao/spinner/helpers.kt]: Okay. I've moved it to `SpinnerAdapterAssertions`.

File: kakao/src/main/kotlin/com/agoda/kakao/spinner/SpinnerAdapterAssertions.kt [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+    fun hasText(text: String) {
+        view.check(ViewAssertions.matches(withSpinnerText(containsString(text))))
+    }
+
+    /**
+     * Check root view is popup and has a DropDownView Descendant
+     *
+     * @return Matcher<Root>
+     */
+    fun isSpinnerPopup() : Matcher<Root> {
```
[the remark this answers is not in the document]
Comment (michaelbukachi) [reply kakao/src/main/kotlin/com/agoda/kakao/spinner/SpinnerAdapterAssertions.kt]: @Unlimity I've moved the matcher to `com.agoda.kakao.common.matchers`

**Comment by Vacxe** [comment]
Thanks for your contribution!

**Comment by Vacxe** [comment]
Close: #118

**Comment by Vacxe** [comment]
@michaelbukachi will merge it in the evening today. Cheers.

**Pull request merged**